src/python/pants/option/options_fingerprinter.py

```diff
@@ -16,7 +16,24 @@
                                        target_option)
 
 
+# TODO: this class's methods could be wrapped in @memoized_method -- we would have to manage the
+# cache size somehow if this is done or it might blow up.
 class Encoder(json.JSONEncoder):
+  def _encode_dict_wrapping_keys(self, d):
+    return {self.encode(k):v for k, v in d.items()}
+
+  def encode(self, o):
+    if isinstance(o, (tuple, set)):
+      o = list(o)
+    elif isinstance(o, dict):
+      o = self._encode_dict_wrapping_keys(o)
+    elif len(o) == 1:
```

Comment:
The first two cases look ok, but this last one seems potentially error prone.

Are there guarantees that the things coming in to this method are always collections? Because otherwise they might not have a `len` method.

Reply:
In the `fingerprint` method in `options_fingerprinter.py`, every input is wrapped in a list if it's not already. This implementation assumes that -- John left some comments below on how to make this more general, which I think addresses your concern specifically, but I think the correct solution is to not wrap everything in a list in `fingerprint`.